### ISSUE TITLE
[bug] ``.. only:: <expr>`` directive content is parsed even if ``<expr>`` evaluates to ``False``

### ISSUE THREAD

**Issue by maxnoe** [sphinx-doc/sphinx#11768]
### Describe the bug

The `toc-tree` directive with the `:glob:` option warns about no matches being there even if the whole statement is excluded from the build using the `.. only:` directive. 

### How to Reproduce

```
sphinx-quickstart minimal_example
cd minimal_example
```

index.rst:

```
.. minimal_glob_only_example documentation master file, created by
   sphinx-quickstart on Fri Nov 24 12:16:13 2023.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

Welcome to minimal_glob_only_example's documentation!
=====================================================

.. only:: non_existent_tag

   IF

   .. toctree::
      :glob:

       foo/*

.. only:: not non_existent_tag

   ELSE
```

Then run `make html`, there is a warning:

```
/home/mnoethe/Projects/sphinx_minimal/index.rst:13: WARNING: toctree glob pattern ' foo/*' didn't match any documents
```

Although the content is excluded with the only tag

### Environment Information

```text
Platform:              linux; (Linux-5.14.0-1051-oem-x86_64-with-glibc2.35)
Python version:        3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:35) [GCC 12.3.0])
Python implementation: CPython
Sphinx version:        7.2.6
Docutils version:      0.20.1
Jinja2 version:        3.1.2
Pygments version:      2.17.2
```

**Comment by picnixz** [comment]
Unfortunately, it appears that we still generate the content and only *after* do we check whether to include or not. I don't know why this was done like that.

The issue is not limited to `glob` or anything. Just be a random <code>:ref:`whatever`</code> and it will still be a warning.

**Comment by pcgaustad** [comment]
This also makes it impossible to have separate TOCs for different targets without having to edit `index.rst` depending on the build target.

My use case is to have multiple TOCs for HTML output, each representing a part (as in the level above chapters) with an item (rst file) for each chapter, while having a separate TOC for LaTeX output where each part is represented by a rst file that includes each chapter with an "include" directive. These rst files are not in HTML builds as that would lead very large pages (containing all the chapters of a part).

**Comment by picnixz** [comment]
Yes but I don't have a good idea on how to correctly solve this one. This may require two passes and the builds would be veeeeery long. 

I should have a look at how GCC handles #if blocks since it might give me some idea. However I won't have time for a few weeks.

By the way, any help for that is welcome.
